homework/index.js

```diff
@@ -29,23 +31,117 @@
     return elem;
   }
 
-  function renderRepoDetails(repo, ul) {
-    createAndAppend('li', ul, { text: repo.name });
+  function addRow(table, labelText, value) {
+    const tr = createAndAppend('tr', table);
+    createAndAppend('th', tr, { text: `${labelText}:` });
+    createAndAppend('td', tr, { text: value });
+    return tr;
   }
 
-  function main(url) {
-    fetchJSON(url, (err, repos) => {
-      const root = document.getElementById('root');
-      if (err) {
+  function renderRepoDetails(repo, div) {
+    div.innerHTML = ''; // to clear result of the previous selection
+    const table = createAndAppend('table', div, { class: 'repos-table' });
+    const firstRow = addRow(table, 'Name', '');
+    createAndAppend('a', firstRow.lastChild, {
+      text: repo.name,
+      href: repo.html_url,
+      target: '_blank',
+    });
+    addRow(table, 'Description', repo.description);
+    addRow(table, 'Forks', repo.forks);
+    addRow(
+      table,
+      'Updated at',
+      new Date(repo.updated_at).toLocaleString('en-GB', { timeZone: 'UTC' }),
+    );
+  }
+
+  /* must declare variable: root, before the next function. */
+  const root = document.getElementById('root');
+
+  function renderContributors(repo, ul) {
+    fetchJSON(repo.contributors_url)
+      .then(contributors => {
+        ul.innerHTML = '';
+        createAndAppend('p', ul, {
+          text: `Contributors:`,
+          id: 'contributors-title',
+        });
+        contributors.forEach(contributor => {
+          const li = createAndAppend('li', ul);
+          const table = createAndAppend('table', li, {
+            class: 'contributor-details',
+          });
+          const a = createAndAppend('a', table, {
+            href: contributor.html_url,
+            target: '_blank',
+          });
+          const tableRow = createAndAppend('tr', a, { class: 'table-row' });
+          createAndAppend('img', tableRow, { src: contributor.avatar_url });
+          createAndAppend('td', tableRow, {
+            text: contributor.login,
+            class: 'user-name',
+          });
+          createAndAppend('td', tableRow, {
+            text: contributor.contributions,
+            class: 'contributions-count',
+          });
+        });
+      })
+      .catch(err => {
         createAndAppend('div', root, {
           text: err.message,
           class: 'alert-error',
         });
-        return;
-      }
-      const ul = createAndAppend('ul', root);
-      repos.forEach(repo => renderRepoDetails(repo, ul));
+      });
+  }
+
+  function createAndSortElements(repos, select) {
+    repos
+      .sort((a, b) => a.name.localeCompare(b.name))
```

Comment:
This function produces a side-effect because it changes one of the parameters that was passed to it: `.sort()` sorts "in place", i.e. it mutates the array you are sorting. What's more, the caller of this function is actually relying on this side-effect, because in `main()` you expect the first element to be in the sorted order. Side-effects like this make your code harder to understand. I would advise doing the sorting in your main function:

```js
fetchJSON(url)
  .then(repos => repos.sort((a, b) => a.name.localeCompare(b.name)))
  .then(repos => {
    createAndSortElements(repos, select);
    renderRepoDetails(repos[0], reposContainer);
    renderContributors(repos[0], ul);
```

Of course, your function name `createAndSortElements` is then no longer applicable. You could rename it to, say, `renderOptionElements()`.

Reply:
Hi Rabie. this is one.

Reply:
I forgot to remove the sorting from renderOptionsElements function. It has been already used in main function.